[5기 - 김현우] SpringBoot Part1 Weekly Mission 제출합니다.

ASak1104 opened the pull request:
## **👩‍💻요구 사항과 구현 내용**

**바우처 관리 Command-line Application**

- [x] **create** 커맨드 지원
    - 생성 가능한 바우처 리스트 (FixedAmountVoucher, PercentDiscountVoucher)
- [x] **list** 커맨드 지원
    - 생성된 바우처 리스트 조회
- [x] **exit** 커맨드 지원
    - 프로그램 종료
- [x] **blacklist** 커맨드 지원
    - 블랙 리스트 명단 조회
    - 명단은 CSV 파일로 부터 가져옴
- [x]  에러 로그는 파일로 기록
- [x]  바우처 정보를 파일로 기록

## **구현 내용**

![diagram](https://github.com/prgrms-be-devcourse/springboot-basic/assets/31983961/eae69288-10eb-42ba-8b77-8add7a33de35)

**VoucherManagementSystemApplication**

- VoucherApplication 실행

**VoucherApplication**

- `Console` 클래스를 이용해 사용자 입력을 받고 적절한 Service를 호출

**CommandType**

- 사용자가 이용할 수 있는 기능을 모아둔 `Enum`

**Console**

- 콘솔 입출력 관련 유틸리티 클래스

**VoucherService**

- 바우처 관련 비즈니스 로직 수행
- 비즈니스 로직에 맞게 적절한 `Repository` 메서드 호출

**BlacklistService**

- 블랙리스트 관련 비즈니스 로직 수행
- `BlacklistRepository`에게 요청을 전달

**VoucherRepository**

- 저장소(파일, 메모리 등)에서 바우처 데이터를 가져와 서비스에게 제공해 줄 기능들을 명시해 놓은 인터페이스

**MemoryVoucherRepository**

- 메모리로 바우처 정보를 관리하는 `VoucherRepository` 구현 클래스

**JSONFileVoucherRepository**

- voucher_data.json 파일에서 바우처를 관리하는 `VoucherRepository` 구현 클래스

**LoadedJSONVoucher**

- JSON으로 저장된 바우처 정보를 매핑하는 클래스
- `VoucherType` 필드에 맞는 바우처 인스턴스를 반환

**Voucher**

- 바우처 정보를 가지고 있는 모델 인터페이스

**VoucherType**

- 바우처 종류를 모아둔 Enum

**FixedAmountVoucher**

- 고정 금액 할인 바우처를 위한  `Voucher` 구현 클래스

**PercentDiscountVoucher**

- 비율(%) 할인 바우처를 위한  `Voucher` 구현 클래스

**User**

- 사용자 정보를 가지고 있는 모델 클래스

## **✅ PR 포인트 & 궁금한 점**

- 프로젝트를 진행하면서 클래스와 메서스, 변수명을 정하는 것이 고민이었습니다.
    
    ex) `LoadedJSONVoucher`, `VoucherManagementSystemApplication`, `VoucherApplication`, `selectCommand`(메서드명)
    
    🙋🏻‍♂️ 현재 네이밍이 적절하게 되어있는지, 그리고 추천하는 네이밍 방식이 있을지 궁금합니다!
    
- `VoucherApplication` 의 실행 방식으로 
(1) 재귀적인 메서드 호출 (현재)
(2) while 문으로 루프를 도는 것
을 고민했는데, 어떤 방식을 사용하는 것이 좋았을까요?
- 현재 저희 어플리케이션에서는 바우처 종류에 상관없이 바우처 정보를 하나의 파일에 저장하고 있습니다. 그렇기 때문에 `JSONFileVoucherRepository` 에서는 바우처 데이터를 불러올 때 모든 종류의 바우처 정보를 가지고 있는 `LoadedJSONVoucher` 클래스로 매핑해서 가져오는데 이 부분을 어떻게 개선할 수 있을지 궁금합니다!
    
    ❗ (문제점) 바우처 종류가 늘어날수록 `LoadedJSONVoucher` 클래스가 갖고 있어야 할 필드가 늘어남
    
    🙋🏻‍♂️ 바우처 종류에 따라 파일을 분리해서 저장하는 것이 좋을까요?
    
    🙋🏻‍♂️ 비슷하게 만약 RDB로 관리한다면 모든 종류의 바우처를 하나의 테이블로 관리하는게 좋은 방식일까요? 
    
- 바우처 정보를 파일에  언제 저장하면 좋을지 궁금합니다!
(1) 프로그램이 종료될 때 한번에 저장 (현재)
(2) 바우처 Create 요청이 올 때마다 저장
(3) 적절한 주기(ex. 1시간, 1일 간격)에 저장
- 블랙리스트 관련 비즈니스 로직을 `UserService`에 넣는 것이 좋을지, 아니면 별도의 `BlacklistService`(현재)에 넣는 것이 좋을지 궁금합니다!

uijin-j committed:
chore: 프로젝트 설정

uijin-j committed:
chore: 프로젝트 구조 수정

uijin-j committed:
docs: gitignore 파일 추가

uijin-j committed:
feat: 명령어 리스트 출력 기능

ASak1104 committed:
style: IO 에러 메시지 수정

ASak1104 committed:
feat: 명령어 선택 기능 추가

uijin-j committed:
feat: VoucherService 생성

uijin-j committed:
feat: 에러 메시지 Console에 모아서 출력

ASak1104 committed:
feat: Voucher 클래스 생성

ASak1104 committed:
feat: VoucherService에 Voucher 생성 기능 추가

ASak1104 committed:
fix: 프로그램이 종료되지 않는 버그 수정

ASak1104 committed:
refactor: 프로젝트 구조 세분화

uijin-j committed:
feat: 바우처 생성 기능 구현

uijin-j committed:
feat: 쿠폰 리스트 조회 기능

uijin-j committed:
feat: 메시지 한글로 변경

ASak1104 committed:
style: Console 모듈을 사용하지 않는 출력 코드 수정

################
# <타입> : <제목> 의 형식으로 제목을 아래 공백줄에 작성
# 제목은 50자 이내 / 변경사항이 "무엇"인지 명확히 작성 / 끝에 마침표 금지
# 예) feat : 로그인 기능 추가

# 바로 아래 공백은 지우지 마세요 (제목과 본문의 분리를 위함)

################
# 본문(구체적인 내용)을 아랫줄에 작성
# 여러 줄의 메시지를 작성할 땐 "-"로 구분 (한 줄은 72자 이내)

################
# 꼬릿말(footer)을 아랫줄에 작성 (현재 커밋과 관련된 이슈 번호 추가 등)
# 예) Close #7

################
# feat : 새로운 기능 추가
# fix : 버그 수정
# docs : 문서 수정
# test : 테스트 코드 추가
# refact : 코드 리팩토링
# style : 코드 의미에 영향을 주지 않는 변경사항
# chore : 빌드 부분 혹은 패키지 매니저 수정사항
################

ASak1104 committed:
feat: 쿠폰 생성 시 유효성 검사 기능 추가

################
# <타입> : <제목> 의 형식으로 제목을 아래 공백줄에 작성
# 제목은 50자 이내 / 변경사항이 "무엇"인지 명확히 작성 / 끝에 마침표 금지
# 예) feat : 로그인 기능 추가

# 바로 아래 공백은 지우지 마세요 (제목과 본문의 분리를 위함)

################
# 본문(구체적인 내용)을 아랫줄에 작성
# 여러 줄의 메시지를 작성할 땐 "-"로 구분 (한 줄은 72자 이내)

################
# 꼬릿말(footer)을 아랫줄에 작성 (현재 커밋과 관련된 이슈 번호 추가 등)
# 예) Close #7

################
# feat : 새로운 기능 추가
# fix : 버그 수정
# docs : 문서 수정
# test : 테스트 코드 추가
# refact : 코드 리팩토링
# style : 코드 의미에 영향을 주지 않는 변경사항
# chore : 빌드 부분 혹은 패키지 매니저 수정사항
################

uijin-j committed:
feat: 로그 기능 추가

ASak1104 committed:
feat: 에러 로그 파일 저장 기능 추가

ASak1104 committed:
refactor: run() 메서드와 사용자 입력 부분 분리

ASak1104 committed:
chore: jackson dependency 추가

ASak1104 committed:
feat: console log appender 추가

ASak1104 committed:
feat: FileManager 모듈 구현

ASak1104 committed:
feat: VoucherType Enum 추가

Voucher 필드 게터 추가

ASak1104 committed:
feat: voucher 데이터를 파일에 저장하고 불러오는 기능 추가

uijin-j committed:
feat: 블랙리스트 조회 기능 추가

ASak1104 committed:
refactor: JSONFileManager를 Repository와 병합

uijin-j committed:
feat: 프로파일 설정 적용

uijin-j committed:
feat: 개발환경에 따라 로깅을 다르게 설정

uijin-j committed:
fix: voucher_data.json 파일이 존재하지 않는 경우 프로그램이 종료되는 오류 해결

uijin-j committed:
refactor: 콘솔과 어플리케이션 메시지 분리

ASak1104 committed:
feat: 애플리케이션 배너 설정

ASak1104 committed:
feat: voucher 목록이 있을 때만 출력하도록 수정

ASak1104 committed:
refactor: voucher 타입에 따라 데이터 저장 형식을 분리

기존 data 필드로 통합 -> amount, percent 로 분리

ASak1104 committed:
style: reformatting code style

ASak1104 committed:
refactor: voucher 서비스와 blacklist 서비스 분리

twotwobread commented on src/main/java/team/marco/vouchermanagementsystem/util/Console.java:
```diff
@@ -0,0 +1,48 @@
+        return Integer.parseInt(readString());
+    }
+
+    public static int readInt(String prompt) {
```
함수명이 readInt인데 인자로 string이 들어가서 인자로 들어간 string을 int로 읽는건가? 생각하게 됩니다. prompt를 출력하는 함수랑 readInt 함수랑 나누는게 더 가독성이 좋을 것 같습니다.

ASak1104 replied:
`readint(String prompt)` 라는 메서드가 말씀해주신대로 혼동될 수 있는 것 같습니다.
`readint()` 메서드가 이미 존재하기 때문에 해당 메서드를 사용하는 부분을 `print()`와 `readint()`로 나누는 방식으로 수정하겠습니다.

commit 73ad52f2f510a69a41a2c649ef0e243c2de4cf37

twotwobread commented on src/main/java/team/marco/vouchermanagementsystem/VoucherApplication.java:
```diff
@@ -0,0 +1,123 @@
+        printList(blacklistService.getBlacklist());
+    }
+
+    private void printList(List<String> list) {
```
출력 기능은 Console이라는 객체가 있으니까 그 쪽으로 옮기는게 좋을 것 같습니다.

ASak1104 replied:
해당 메서드가 `List` 자료형을 출력하는 목적에서 `Console` 유틸에서 처리하는 것도 좋은 것 같습니다.
하지만 제가 생각했을 때 해당 메서드가 `Console.print("조회가 완료되었습니다.");` 와 같이 `Console` 유틸과는 관계없는 메시지를 포함하고 `join`한 문자열이 공백문자일 경우 출력을 하지 않는 기능을 한다는 것은 `Console` 객체의 역할과는 거리가 있는 것 같습니다.

그래서 해당 메서드가 조회된 결과를 출력한다는 의미에서 아래와 같이 수정해봤습니다.

> 조회 결과를 출력한다는 의미에서 메서드 명과 해당 파라미터 명을 변경했습니다.
> printList() -> printInquireResult()
> list -> InquiredList
> commit https://github.com/prgrms-be-devcourse/springboot-basic/commit/a6a8b3f7bf66800608344101ed59fe806f0b8832

서브 멘토님께서는 어떻게 생각하시나요?

twotwobread commented on src/main/java/team/marco/vouchermanagementsystem/VoucherApplication.java:
```diff
@@ -0,0 +1,123 @@
+            Console.print(e.getMessage());
+        }
+
+        selectCommand();
```
함수를 호출하면 stack 메모리를 소모하게 되고 해당 재귀함수는 중단하는 로직도 없기 때문에 depth가 깊어지다가 stack over flow가 발생하지 않을까 생각합니다. 그리고 while문이 이 로직에서 더 적합하고 가독성도 코드를 이해하기도 쉬울 것 같습니다.

ASak1104 replied:
Java의 경우 TCO (Tail-Call Optimization)를 지원하지 않기 때문에 서브 멘토님께서 지적해주신 부분을 고려하는 것이 타당한 것 같습니다. 말씀해주신 while 문 방식으로 로직을 변경해보도록 하겠습니다!

commit 97f1d77975b1598c572399f04ab0513ffb7f248e

twotwobread commented on src/main/java/team/marco/vouchermanagementsystem/VoucherApplication.java:
```diff
@@ -0,0 +1,123 @@
+                    return;
+                }
+            }
+        } catch (NumberFormatException e) {
```
로직과 예외 핸들링이 섞이면 로직에 대한 이해가 어려워집니다. 예외 핸들링과 로직을 분리해보면 좋을 것 같습니다.

ASak1104 replied:
commit https://github.com/prgrms-be-devcourse/springboot-basic/commit/97f1d77975b1598c572399f04ab0513ffb7f248e

twotwobread commented on src/main/java/team/marco/vouchermanagementsystem/model/Voucher.java:
```diff
@@ -0,0 +1,11 @@
+
+import java.util.UUID;
+
+public interface Voucher {
```
메서드의 추상화보다 인스턴스 변수나 Object 클래스 메서드 오버라이딩이 더 적합할 것 같아서 추상화 클래스로 만드는게 지금 상황에선 나아보입니다

ksy90101 replied:
수영님 저와 마음이 통했넹.

ASak1104 replied:
페어프로그래밍 과정에서 계속해서 고민한 부분인데요. 인터페이스를 사용한 이유는 크게 없습니다.
왜 인터페이스를 사용했냐는 질문에 대한 제 생각을 말씀드리자면 리더님의 강의에서 인터페이스로 구현하는 방식을 사용했다
그래서 자연스럽게 인터페이스를 활용하는 방식으로 구현한 것 같습니다.
서브 멘토님과 멘토님께서 말씀해주신대로 수정해봤습니다!

commit 29c3e8cb120d73256faee969528de231bef94d43

twotwobread commented on src/main/java/team/marco/vouchermanagementsystem/model/PercentDiscountVoucher.java:
```diff
@@ -0,0 +1,58 @@
+
+    @Override
+    public String getInfo() {
+        return format("{0}% 할인 쿠폰", percent);
```
toString을 이용하는게 낫지 않을까요?

ASak1104 replied:
확실히 `getInfo()`라는 메서드를 추가로 선언하는 것 보다 `toString()` 메서드를 오버라이딩하는 것으로 얻을 수 있는 이점이 있다고 생각합니다.

저는 `toString()` 메서드의 목적이 객체의 상태 표현이라고 생각하는데요. 그래서 `toString()` 메서드가 단순히 애플리케이션에서 필요한 정보만이 아닌 debug 환경에서의 상태 표현처럼 다른 방면에서 활용된 측면이 있다고 생각합니다.
그래서 _애플리케이션에서 사용하는 정보_ 라는 의미에서 `getInfo()` 라는 메서드를 사용해서 구현해 봤습니다.

저는 이렇게 생각해봤는데 서브 멘토님께서는 어떻게 생각하시나요?
정도가 지나친 고민이었다면 다시 수정해보도록 하겠습니다!!

twotwobread commented on src/main/java/team/marco/vouchermanagementsystem/model/User.java:
```diff
@@ -0,0 +1,25 @@
+
+    public String getInfo() {
+
+        return format("id: {0}, 고객명: {1} ", id, name);
```
toString으로 표현하는게 나을 것 같습니다.

twotwobread commented on src/main/java/team/marco/vouchermanagementsystem/repository/BlacklistRepository.java:
```diff
@@ -0,0 +1,42 @@
+                    .map(s -> s.split("[;,]"))
+                    .forEach(data -> loaded.add(new User(UUID.fromString(data[0]), data[1])));
+        } catch (IOException e) {
+            throw new RuntimeException("데이터를 가져올 수 없습니다.");
```
RuntimeException은 너무 넓은 exception입니다. 좀 더 의미를 전달할 수 있는 exception을 사용하면 좋을 것 같습니다.

ASak1104 replied:
말씀해주신대로 수정했습니다!

commit 4b22fafb34ca22035b56c07eba0f026053a34f90

여기서 한 가지 궁금한 점이 생겼는데
매우 구체적인 예외를 던질 수록 이를 상위 레이어의 모듈에서 처리할 필요가 생길 것 같은데 이럴 때 예외의 구체성을 나누는 기준을 어떻게 잡을 수 있을까요?

예를 들어 현재 애플리케이션은 `NumberFormatException`과 `IllegalArgumentException`을 제외한 모든 경우는 `run()` 메서드에서 프로그램이 종료되도록 처리되지만 만약 `JSONFileSaveException`, `CSVFileLoadException` 등등 다양한 예외를 던지고 이에 대한 모든 예외처리를 해야한다면 정말 좋지 못한 것 같은데 어떤 예외를 상속하는 구체적인 예외를 던지고 비교적 추상적인 예외를 핸들링 해야할까요? 그렇다면 커스텀 예외를 던지기 위해 커스텀 예외에 의존하게 되면서 변경점에 노출될 수도 있지 않을까요?

`VoucherApplication`은 서비스가 정확히 어떤 `Repository`를 사용하고 어떤 `exception`을 던지는지 모를 필요가 있고 이러한 점이 `UncheckedException`을 던지는 이유 중 하나이니까요.

twotwobread commented on src/main/java/team/marco/vouchermanagementsystem/repository/JSONFileVoucherRepository.java:
```diff
@@ -0,0 +1,71 @@
+        try {
+            loadedList = objectReader.readValue(file);
+        } catch (IOException e) {
+            throw new RuntimeException("파일을 불러오는 과정에서 오류가 발생했습니다.");
```
여기도 더 의미있는 예외를 사용하는게 좋을 것 같습니다.

ASak1104 replied:
commit https://github.com/prgrms-be-devcourse/springboot-basic/commit/4b22fafb34ca22035b56c07eba0f026053a34f90

twotwobread commented on src/main/java/team/marco/vouchermanagementsystem/model/FixedAmountVoucher.java:
```diff
@@ -0,0 +1,50 @@
+    private final int amount;
+
+    public FixedAmountVoucher(int amount) {
+        if (amount <= 0) {
```
매직 넘버를 사용하기 보단 상수 바꿔서 사용하면 좋을 거 같습니다.

ASak1104 replied:
commit c5a9e7e7563ee4a22a9980e7bda0033d8e85b52b

twotwobread commented on src/main/java/team/marco/vouchermanagementsystem/model/PercentDiscountVoucher.java:
```diff
@@ -0,0 +1,58 @@
+    }
+
+    private void validate(int percent) {
+        if (percent <= 0) {
```
여기도 매직넘버보단 상수를 이용하면 좋을 것 같습니다.

ASak1104 replied:
commit c5a9e7e7563ee4a22a9980e7bda0033d8e85b52b

twotwobread commented on src/main/java/team/marco/vouchermanagementsystem/model/PercentDiscountVoucher.java:
```diff
@@ -0,0 +1,58 @@
+        return VoucherType.PERCENT;
+    }
+
+    public int getPercent() {
```
사용되지 않는 메서드입니다. 삭제하는 것이 좋을 것 같습니다.

ASak1104 replied:
해당 메서드는 코드 상에서 직접적으로 사용되지 않지만 JSON 파일로 데이터를 저장할 때 ObjectMapper를 통해 간접적을 호출됩니다.
데이터 저장을 위해 해당 메서드가 존재하는데 이러한 방식은 잘못된 것일까요?
수정을 한다면 어떤 방식으로 해야되는지 궁금합니다!

ksy90101 commented on src/main/java/team/marco/vouchermanagementsystem/repository/BlacklistRepository.java:
```diff
@@ -0,0 +1,42 @@
+        try (BufferedReader reader = Files.newBufferedReader(Paths.get(path), StandardCharsets.UTF_8)) {
+            reader.readLine(); // skip header
+            reader.lines()
+                    .map(s -> s.split("[;,]"))
```
[;.]를 상수화를 해서 클린코드를 지켜보면 어떨까요?

ASak1104 replied:
commit a72f674f982c6e454fea8b77166b9ee13c594f8a

ksy90101 commented on src/main/java/team/marco/vouchermanagementsystem/model/PercentDiscountVoucher.java:
```diff
@@ -0,0 +1,58 @@
+import static java.text.MessageFormat.format;
+
+public class PercentDiscountVoucher implements Voucher {
+    private static final Logger logger = LoggerFactory.getLogger(FixedAmountVoucher.class);
```
로그 클래스가 다르네요?

ASak1104 replied:
이런 실수를... 감사합니다!

commit d7cb44776f9ff8cb36b574edaa8ffd00c8a673b1

ksy90101 commented on src/main/java/team/marco/vouchermanagementsystem/VoucherManagementSystemApplication.java:
```diff
@@ -0,0 +1,24 @@
+import org.springframework.boot.SpringApplication;
+import org.springframework.boot.autoconfigure.SpringBootApplication;
+import org.springframework.context.ApplicationContext;
+import team.marco.vouchermanagementsystem.model.VoucherType;
```
사용하지 않는 import
<img width="971" alt="image" src="https://github.com/prgrms-be-devcourse/springboot-basic/assets/53366407/de9b1b53-b719-4fa7-a4a0-7865a6a8d002">
인텔리제이 설저엥 이러한 옵션창이 있는데 이러한 옵션창을 이용해 저장시 리팩토링하게 한다면 이러한 문제는 쉽게 해결될거 같아요

ASak1104 replied:
IDE 설정까지 알려주셔서 감사합니다!!

commit c262bd45d52d9df898682958e781f15ab49e9f88

ksy90101 commented on src/main/java/team/marco/vouchermanagementsystem/VoucherManagementSystemApplication.java:
```diff
@@ -0,0 +1,24 @@
+    public static void main(String[] args) {
+        ApplicationContext context = SpringApplication.run(VoucherManagementSystemApplication.class, args);
+
+        logger.info("Program start " + context.getEnvironment().getActiveProfiles()[0]);
```
logger.info("Program start {0}", context.getEnvironment().getActiveProfiles()[0]);

이런식으로 format형식으로 쓰면, 지금은 0이지만, 타입까지 지정하면 좋겠죠. 좀더 좋습니닷!

성능상으로도요 ㅎㅎ

ASak1104 replied:
commit a389446ef6cebc13bb8eee0461f30827afe96aba

ksy90101 commented on src/main/java/team/marco/vouchermanagementsystem/CommandType.java:
```diff
@@ -0,0 +1,15 @@
+package team.marco.vouchermanagementsystem;
```
이것또한 패키지로 묶을수 잇을거 같은데요?
패키지화를 통해 클린코드를 지켜보면 어떨까요?

ASak1104 replied:
`application` 이라는 패키지로 VoucherApplication과 CommandType 이넘을 묶어봤습니다.

commit 4e677b96dbed74a78d137434917aa44fc7213a04

만약 `WebApplication` 이라는 클래스가 추가된다면 `VoucherApplication` 이라는 `Interface`를 도입하는 것도 좋아보이는데
이것에 대해선 어떻게 생각하시나요? 멀티쓰레드 환경은 아니지만 `Runnable` 인터페이스를 사용하는게 좋을까요?

ksy90101 commented on src/main/java/team/marco/vouchermanagementsystem/VoucherApplication.java:
```diff
@@ -0,0 +1,123 @@
+    }
+
+    public void selectCommand() {
+        Console.print("""
+                === 쿠폰 관리 프로그램 ===
+                exit: 프로그램 종료
+                create: 쿠폰 생성
+                list: 쿠폰 목록 조회
+                blacklist: 블랙 리스트 유저 조회""");
```
차후 view 코드가 변경되어야 한다면 어떻게 될까요?
지금은 View와 Application과 너무 강하게 엮여 있는거 같기도 하네요.
사실 Application보단 Controller의 역할이 지금은 여기가 더 큰게 아닐까 해요.
Controller를 찾아보고 Controller패턴으로 한번 적용해보면 어떨까 해요.

ASak1104 replied:
해당 애플리케이션이 Web이 아니다 보니 `Controller`가 필요한지 필요하다면 어떤 방향으로 수정하는 것이 좋을 지 궁금합니다

현재 `VoucherApplication` (현 `ConsoleApplication`) 클래스는 콘솔을 통해 유저에게 `View`를 전달하고 입력을 받는데요.
그렇다면 `Controller`가 하는 역할은 현재 코드에서 서비스를 호출하는 부분을 단순히 2중으로 연결하는 역할이 아닌가하는 생각이 들었습니다.

개발 초기부터 `Controller` 도입을 고려했지만 앞선 질문에 대한 명쾌한 답을 내지 못해서 도입하지 않았던 것 같습니다.
Web MVC에서는 URI를 통해 메서드와 서비스에 대한 분기가 나누어지지만 해당 애플리케이션은 콘솔을 사용하다보니 `Controller`가 `Console`에 의존하지 않으면 이러한 역할을 하는 것이 불가능하니까요.

그래서 `Console` 유틸을 `Controller`가 사용한다면 `View`와 `Controller`의 분리가 가능하기에 이런 방식을 말씀해주신건지 궁금합니다!

만약 맞다면 `ConsoleApplication`을 `View`와 `Controller`로 분리해서 역할을 명확히 하도록 수정해보겠습니다!

ksy90101 commented on src/main/java/team/marco/vouchermanagementsystem/VoucherApplication.java:
```diff
@@ -0,0 +1,123 @@
+    }
+
+    private void createVoucher() {
+        logger.info("Call createVoucher()");
```
일반 사용자들이 사용할때 logger가 계속 찍히네요.
콘솔로 실행해서 실행할때도 실제 어플리케이션이에요.
Logger는 실제 사용시에는 안나오게 하면 어떨까 생긱이 드네요.

ASak1104 replied:
초기에는 dev 환경에서 info 레벨을 로깅하고
console 환경은 error 파일만 로깅하고자 했지만 로거 설정에 실수가 있었습니다!
감사합니다!!

commit 8521b761ff2dee4c746474d228f99676bee7665d

ksy90101 commented on src/main/java/team/marco/vouchermanagementsystem/VoucherApplication.java:
```diff
@@ -0,0 +1,123 @@
+    }
+
+    private void createPercentDiscountVoucher() {
+        logger.info("Call createPercentDiscountVoucher()");
```
그리고 이건 중복코드네요.
차후 시간이 난다면 이런걸 AOP로 적용해보면 어떨까 생각이 드네요.

ASak1104 replied:
저도 그렇게 생각합니다!
2차 과제 기간 동안 적용해보도록 노력하겠습니다!

ksy90101 commented on src/main/java/team/marco/vouchermanagementsystem/VoucherApplication.java:
```diff
@@ -0,0 +1,123 @@
+        printList(blacklistService.getBlacklist());
+    }
+
+    private void printList(List<String> list) {
```
list라는 단어는 추상적인 단어에요.
의미있는 변수명을 써주세요.

ASak1104 replied:
조회 결과를 출력한다는 의미에서 메서드 명과 해당 파라미터 명을 변경했습니다.
- `printList()` -> `printInquireResult()`
- `list` -> `InquiredList`

commit a6a8b3f7bf66800608344101ed59fe806f0b8832

ksy90101 commented on src/main/java/team/marco/vouchermanagementsystem/model/FixedAmountVoucher.java:
```diff
@@ -0,0 +1,50 @@
+public class FixedAmountVoucher implements Voucher {
+    private static final Logger logger = LoggerFactory.getLogger(FixedAmountVoucher.class);
+
+    private final UUID id;
+    private final int amount;
```
(질문)
해당 모델들의 validate는 따로 생각해본적이 없나요?
예를들어 int범위 이상이 된다면 어떻게 될까요?

이럴때 좋은 건 Spring Validation 라이브러리죠^^

ASak1104 replied:
validate를 생각해서 `validate()` 메서드를 사용했습니다. `FixedAmountVoucher`는 `PercentAmountVoucher`와 달리 최소값만 판단한다는 이유로 페어프로그래밍 시 따로 `validate()` 메서드를 사용하지 않고 constructor에서 데이터에 대한 검증을 했습니다. (현재는 `validate()` 메서드를 사용합니다. commit https://github.com/prgrms-be-devcourse/springboot-basic/commit/c5a9e7e7563ee4a22a9980e7bda0033d8e85b52b)

말씀해주신 jakarta.validation 과 spring-boot-starter-validation 라이브러리를 활용해보려 했지만 여러 문제점이 생겼습니다.

1. @ Valid (@ Validated 포함)는 Controller 빈에서 사용이 가능하다
2. 현재 애플리케이션은 따로 DTO를 사용하지 않는다 (생성을 위해 primitive int 하나만 필요) 따라서 스프링 Web처럼 Controller에서 @ Valid Request param 같은 검증은 사용할 수 없다.
3. jakarta.validation의 Specification 구현체인 [Hibernate Validator](https://docs.jboss.org/hibernate/stable/validator/reference/en-US/html_single/#validator-gettingstarted-createproject) 를 사용하면 service 단에서 validation을 할 수 있지만 Validator를 composition으로 사용하는 방식이 과연 좋은 코드인지 알 수 없다.
    ```java
    @Service
    public class VoucherService {
        private final VoucherRepository voucherRepository;

        private final Validator validator;

        public VoucherService(VoucherRepository voucherRepository) {
            this.voucherRepository = voucherRepository;

            ValidatorFactory validatorFactory = Validation.buildDefaultValidatorFactory();

            validator = validatorFactory.getValidator();
        }

        public void createFixedAmountVoucher(int amount) {
            Voucher voucher = new FixedAmountVoucher(amount);

            Set<ConstraintViolation<Voucher>> violations = validator.validate(voucher);

            if (!violations.isEmpty()) {
                throw new IllegalArgumentException();
            }

            voucherRepository.save(voucher);
        }

    ... 중략 ...
    ```

멘토님께서 어떤 방식으로 validation 라이브러리를 사용하길 바라셨는지 모르겠습니다.
구체적으로 알려주실 수 있으실까요?

우선적으로 고정 할인의 최대값을 설정해봤습니다.

commit 0ea01703a9bddbf1f200f8a74b416ea258f29f2d

ksy90101 commented on src/main/java/team/marco/vouchermanagementsystem/model/FixedAmountVoucher.java:
```diff
@@ -0,0 +1,50 @@
+import static java.text.MessageFormat.format;
+
+public class FixedAmountVoucher implements Voucher {
+    private static final Logger logger = LoggerFactory.getLogger(FixedAmountVoucher.class);
```
모델을 구현하는데 interface가 좋은걸까요?

사실 전 상속 또는 조합 이 두가지 방식이 더 어울린다고 생각하는데요

왜 interface로 구현했는지 궁금해졌네요 :)

ASak1104 replied:
페어프로그래밍 과정에서 계속해서 고민한 부분인데요. 인터페이스를 사용한 이유는 크게 없습니다.
왜 인터페이스를 사용했냐는 질문에 대한 제 생각을 말씀드리자면 _리더님의 강의에서 인터페이스로 구현하는 방식을 사용했다_ 입니다
그래서 자연스럽게 인터페이스를 활용하는 방식으로 구현한 것 같습니다.
서브 멘토님과 멘토님께서 말씀해주신대로 수정해봤습니다!

commit https://github.com/prgrms-be-devcourse/springboot-basic/commit/29c3e8cb120d73256faee969528de231bef94d43

ksy90101 commented on src/main/java/team/marco/vouchermanagementsystem/model/PercentDiscountVoucher.java:
```diff
@@ -0,0 +1,58 @@
+    public PercentDiscountVoucher(int percent) {
+        validate(percent);
+
+        this.id = UUID.randomUUID();
```
이런 로직자체가 상속이였더면 중복코드를 제거할수 있었을꺼 같아요.

ASak1104 replied:
제가 생각한 중복 코드는 uuid 생성을 위한 `UUID.randomUUID()` 메서드가 동일하게 호출되고 `getID()` 메서드를 오버라이딩 한다는 점이어서 상속을 통해 해당 부분을 제거해 봤습니다.

commit https://github.com/prgrms-be-devcourse/springboot-basic/commit/29c3e8cb120d73256faee969528de231bef94d43

그러던 중에 궁금했던 점이`validate(params..)` 메서드 또한 상위 클래스인 `Voucher`로 옮길 필요가 있을까? 였는데요
TemplateMethod 패턴처럼 `hook()` 메서드로 묶는 것이 구현체의 파라미터 개수가 모두 하나라 현재 가능한 방법이고 코드 상에서 반복되는 호출을 부모 클래스에만 정의하면 되니까요
추후에 파라미터 개수가 늘어난 구현체가 생기면 어떨까? 라는 생각을 하면 그렇게 좋아보이진 않아서 고민이되네요.

멘토님께서는 어떻게 생각하시나요?

ksy90101 commented on src/main/java/team/marco/vouchermanagementsystem/model/User.java:
```diff
@@ -0,0 +1,25 @@
+
+public class User {
+    private final UUID id;
+    private final String name;
```
여기도 validate를 고민해보면 어떨까요?

ksy90101 commented on src/main/java/team/marco/vouchermanagementsystem/repository/BlacklistRepository.java:
```diff
@@ -0,0 +1,42 @@
+
+@Repository
+public class BlacklistRepository {
+    private final List<User> blacklist;
```
해당 리스트는 여러 스레드에서도 같은걸 가지고 잇어야 하지 않을까요?
static화를 해야 하지 않을까 생각이 들어서요.

ASak1104 replied:
멘토님 말씀대로 모든 쓰레드가 동일한 리스트를 공유할 필요가 있다고 생각합니다!
그런데 Spring 빈은 prototype과 같은 특수한 라이프 사이클의 빈을 제외하면 싱글톤으로 관리되지 않나요??
그래서 해당 빈이 하나만 존재해서 데이터 리스트와 생성자를 통한 `load()` 메서드 호출이 한 번만 이루어진다고 생각했는데
제가 잘못 생각한 부분이 있을까요?

ksy90101 commented on src/main/java/team/marco/vouchermanagementsystem/repository/BlacklistRepository.java:
```diff
@@ -0,0 +1,42 @@
+public class BlacklistRepository {
+    private final List<User> blacklist;
+
+    public BlacklistRepository(@Value("${file.path.blacklist}") String path) {
```
https://rutgo-letsgo.tistory.com/93

ConfigurationProperties
도 고려해보셨을까요?

ASak1104 replied:
`ConfigurationProperties`가 존재하는지 몰라서 사용을 고려해보질 못 했습니다!
찾아보니 프로퍼티를 관리하는 차원에서 충분히 고려해볼만한 좋은 방법인 것 같습니다!

파일 경로를 `FilePathProperties` record로 관리해봤습니다!

commit 050a33a607ee72498f3d690a17202c5ca48faaf5

ksy90101 commented on src/main/java/team/marco/vouchermanagementsystem/repository/BlacklistRepository.java:
```diff
@@ -0,0 +1,42 @@
+            reader.readLine(); // skip header
+            reader.lines()
+                    .map(s -> s.split("[;,]"))
+                    .forEach(data -> loaded.add(new User(UUID.fromString(data[0]), data[1])));
```
forEach로 add하기보단 map으로 한번에 만드는게 좋지 않을까요?

ASak1104 replied:
그렇게 하면 리스트 원소의 추가 삭제를 확실히 막을 수 있기 때문에 좋은 것 같습니다!
감사합니다!

commit 34b37c773bc35d2cbd13338bcff904b87ef66f30

ksy90101 commented on src/main/java/team/marco/vouchermanagementsystem/repository/BlacklistRepository.java:
```diff
@@ -0,0 +1,42 @@
+    }
+
+    public List<User> findAll() {
+        return blacklist;
```
이렇게 되면, 외부에서 수정하면 blacklist는 어떻게 될까요?
이러한 문제를 해결하기위해 불변으로 해보면 어떨까요?
https://lts0606.tistory.com/133
이런식으로 말이죠.

ASak1104 replied:
이전 커밋에서 `load()` 메서드가 `stream.toList()` 사용하면서 이미 불변 리스트가 반환되지만
`Collections.unmodifiableList()` 내부 로직 상 이미 불변 리스트일 때 그대로 반환해서 리소스가 낭비되지 않고
코드 상에서 이를 명시하기 위한 목적으로 사용할 수 있기에 말씀해주신 방식으로 수정해보겠습니다!

이전 commit https://github.com/prgrms-be-devcourse/springboot-basic/commit/34b37c773bc35d2cbd13338bcff904b87ef66f30

commit 7752fdb3757f4d6e24b5891b82f6de19047ed892

ksy90101 commented on src/main/java/team/marco/vouchermanagementsystem/repository/JSONFileVoucherRepository.java:
```diff
@@ -0,0 +1,71 @@
+@Repository
+public class JSONFileVoucherRepository implements VoucherRepository, DisposableBean {
+    private final Map<UUID, Voucher> voucherMap;
+    private final ObjectMapper objectMapper;
```
👍

ksy90101 commented on src/main/java/team/marco/vouchermanagementsystem/repository/JSONFileVoucherRepository.java:
```diff
@@ -0,0 +1,71 @@
+    }
+
+    @Override
+    public List<Voucher> findAll() {
```
여기도 불변을 한번 고려해보세요 :)

ASak1104 replied:
이전 Blacklist 레포지토리와 다르게 해당 메서드는 코드 상에서
불변 리스트가 반환되는 것이 잘 명시되어 있는 것 같습니다!
`return voucherMap.values().stream().toList()`

한 번 다시 고려해보겠습니다!!

ksy90101 commented on src/main/java/team/marco/vouchermanagementsystem/repository/JSONFileVoucherRepository.java:
```diff
@@ -0,0 +1,71 @@
+        try {
+            objectMapper.writeValue(file, voucherMap.values());
+        } catch (IOException e) {
+            throw new RuntimeException("파일을 저장하는 과정에서 오류가 발생했습니다.");
```
의미잇는 예외 사용해주세요.

ASak1104 replied:
commit 4b22fafb34ca22035b56c07eba0f026053a34f90

ksy90101 commented on src/main/java/team/marco/vouchermanagementsystem/repository/LoadedJSONVoucher.java:
```diff
@@ -0,0 +1,47 @@
+        // for object mapper deserializing
+    }
+
+    public Voucher convertToVoucher() {
+        return switch (getType()) {
+            case FIXED -> new FixedAmountVoucher(getId(), getAmount());
+            case PERCENT -> new PercentDiscountVoucher(getId(), getPercent());
+        };
+    }
```
이렇게 되면 enum도 가서 추가해야하고 여기도 와서 추가해야하고.
이렇게 하기 보단, 차라리 enum에서 전략패턴을 써보면 어떨까요?

ASak1104 replied:
> 이렇게 되면 enum도 가서 추가해야하고 여기도 와서 추가해야하고.
이 부분에 대해선 확실히 개선이 필요한 것 같습니다!

전략 패턴을 어떻게 적용하는지 이해가 잘 되지 않아서
우선적으로 기존 `LoadedJSONVoucher`가 JSON 데이터 형식과 큰 관련이 없기 때문에 `LoadedVoucher`로 클래스 명을 변경하고 `VoucherType` Enum에서 해당 `LoadedVoucher`를 `Voucher` 클래스로 변경하도록 수정했습니다.

composition이나 전략 패턴으로 어떻게 변경하는 것을 생각하셨는지 설명해주시면 다시 수정해보겠습니다!

commit 25aeafb31774fb72c374fb81aadf36dd5aabdc3d

ksy90101 commented on src/main/java/team/marco/vouchermanagementsystem/repository/LoadedJSONVoucher.java:
```diff
@@ -0,0 +1,47 @@
+
+public class LoadedJSONVoucher {
+    private UUID id;
+    private VoucherType type;
+    private String info;
+    private int amount;
+    private int percent;
```
validate를 고민해보세요.

ksy90101 commented on src/main/java/team/marco/vouchermanagementsystem/repository/MemoryVoucherRepository.java:
```diff
@@ -0,0 +1,26 @@
+
+    @Override
+    public List<Voucher> findAll() {
+        return voucherMap.values().stream().toList();
```
스트림은 한줄의 점 하나씩으로 쓰는게 더 좋더라구요~

ASak1104 replied:
한 줄에 써서 읽기 쉬우면 한 줄로 쓰자!
stream() api를 호출한 다음부터 줄을 바꾸자!
등의 규칙으로 적용했었는데
멘토님처럼 스트림과 같은 메소드 체이닝에 대해선 한 줄에 하나의 메서드가 호출하도록 수정해봤습니다!

commit https://github.com/prgrms-be-devcourse/springboot-basic/commit/71a2ae45a1dd4fd6591ab788475aa932bbe5e48e

ksy90101 commented on src/main/java/team/marco/vouchermanagementsystem/util/Console.java:
```diff
@@ -0,0 +1,48 @@
+    }
+
+    public static int readInt() {
+        return Integer.parseInt(readString());
```
에러 터지면 어떻게 되나요?

ASak1104 replied:
`NumberFormatException`이 발생하고 이를 `VoucherApplication` (현 `ConsoleApplication`)이 처리하게 됩니다.

_해당 예외가 터진다는 것을 사용자 입장에서는 예측하기 힘들다_ 라는 이유로
코드 수정이 되면 좋겠다는 의미셨을까요?

ksy90101 reviewed (changes requested):
간단한 피드백을 남겼어요 파이팅입니닷.

ASak1104 committed:
refactor: 혼동을 줄 수 있는 메서드 이름 변경 및 readint(String prompt) 메서드 삭제

ASak1104 committed:
style: printList 메서드 명 변경

메서드 역할을 구체적으로 명시

ASak1104 committed:
refactor: 애플리케이션 실행 방식을 제귀에서 반복으로 변경

ASak1104 committed:
refactor: IOException 래핑 예외를 Runtime에서 UncheckedIO로 변경

ASak1104 committed:
refactor: validation에 사용되는 숫자를 상수로 변경

ASak1104 committed:
refactor: voucher 인터페이스를 추상 클래스로 변경

ASak1104 committed:
refactor: csv 파일 regex를 상수로 변경

ASak1104 committed:
fix: 로거 필드 클래스 부분 수정

ASak1104 committed:
style: 불필요한 import 제거

ASak1104 committed:
refactor: 프로그램 실행 로그 메시지 포맷 지정

ASak1104 committed:
refactor: application 패키지 구성 및 VoucherApplication 이름 변경

기존 상위 모듈인 CommandType 이넘과 
VoucherApplication을 application 패키지로 이동
VoucherApplication 이름을 ConsoleApplication으로 변경

ASak1104 committed:
fix: dev, local profile의 로거 설정 수정

ASak1104 committed:
fix: 명령어 입력 상황에서 잘못된 입력 예외처리가 되지 않는 버그 수정

ASak1104 committed:
fix: local profile의 repository 변경

ASak1104 committed:
fix: FixedAmountVoucher의 최대 할인 금액 설정

최대값: 10억

ASak1104 committed:
refactor: blacklist 데이터 로드 방법을 stream API로 변경

ASak1104 committed:
style: blacklist 필드를 반환할 때 불변성 명시

ASak1104 committed:
style: stream API의 코드 줄바꿈 방식 통일

ASak1104 committed:
style: 사용하지 않는 생성자 제거

ASak1104 committed:
refactor: @Value 어노테이션을 ConfigurationProperties 사용으로 변경

ASak1104 committed:
refactor: 로드된 voucher 데이터를 변환하는 로직을 이넘으로 이동

ASak1104 committed:
style: 로드한 voucher 리스트의 변수명 수정

기존 loadedList -> loadedVouchers

ksy90101 commented on src/main/java/team/marco/vouchermanagementsystem/application/ConsoleApplication.java:
```diff
@@ -0,0 +1,139 @@
+import java.util.List;
+
+@Component
+public class ConsoleApplication {
```
해당 부분에 대해서 고민을 해봤는데,
제가 생각하는 Application이라는 레이어와 현우님이 생각하시는 Application 레이어가 다른거 같습니다.
저는 Application은 main메서드가 존재하고 실행할수 있는 영역이라고 생각했지만, 현우님이 해주신 영역에서는 Application은 이러한 로직을 실행하는 공간같아요. 사실 지금 다시 코드를 보면, 제가 말한 Controller와 여기에 있는 Application과 다른건 없고 이름의 차이만 있다고 생각합니다 :)

ASak1104 replied:
부드럽게 말씀해주셔서 감사합니다!!
1주차 요구 사항을 구현할 때는 제 관점에서 현재 Application의 구조가 이상적인 형태라고 생각했는데요
이번 2주차 요구 사항을 보고 고민을 해보니 현재 구조를 유지하는 것은 확장성과 유지 보수의 관점에서는 좋지 못하다는 생각이 들었습니다.

아마도 2주차는 VoucherController (또는 ConsoleVoucherController), WalletController, BlacklistController로 현재 애플리케이션의 로직이 앞선 3가지 클래스를 통해 분리되고 각 애플리케이션의 기능이 서로 독립되도록 방향으로 구현을 할 것 같습니다.
1주차와 달리 2주차의 요구사항을 만족하기 위해선 이러한 분리가 저에게 객체지향적이고 좋은 코드로서 느껴졌거든요!
멘토님의 피드백이 여러 방면에서 고민할 수 있는 계기가 된 것 같습니다!
2주차 PR에서 뵙겠습니다.

감사합니다!

ksy90101 replied:
2주차에 방향을 보고 더 좋은 방향으로 가봐요~

ksy90101 reviewed (approved):
고생하셨습니다 2단계에서 봐요!
